### PR TITLE
SetCartRuleProductRestrictions CQRS handling

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add cart rule product restrictions handling into SetCartRuleRestrictionsCommand/handler. Add additional behat tests.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yno
| How to test?      | New command covered with behats and unused in any exposed UI page (new cart rules page is not yet finished) so it doesn't need manual testing.
| Fixed ticket?     | part of https://github.com/PrestaShop/PrestaShop/issues/10547
| Related PRs       | If theme, autoupgrade or other module change is needed to make this change work, provide a link to related PRs here.
| Sponsor company   | Your company or customer's name goes here (if applicable).

### CODE REVIEW

File: src/Adapter/CartRule/CommandHandler/SetCartRuleProductRestrictionsHandler.php
Comment (zuk3975) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleProductRestrictionsHandler.php]: as noticed in another PR (https://github.com/PrestaShop/PrestaShop/pull/32483), $cartRule->reduction_product is affected as well (at least in legacy code), but the two features (product restriction in conditions tab & specific product selection in actions tab) seems mixed up together and needs to be sorted out (restrictions defines rules when discount applies, reduction_product defines that the applicable discount should only be applied to that specific product and not whole order, so basically when there is a reduction_product it shouldn't need to be added to restrictions, because its already handled separately in FO.)
Example of current behavior - when you select specific product in actions tab, then that product appears in conditions -> product restrictions :sweat:

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: I'm not sure this should be the responsibility of the command to indicate it's empty 🤔 The command is a pure DTO and `isEmpty` is not an information of a field being edited it's a state of the command itself

It makes more sense for the handler to check this IMHO

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: You're using this as a generic updater but it always modifies `$cartRule->cart_rule_restriction` I think you can keep this one for cart rule restriction only especially since you use it to update all the related cart rules

Besides you already update `$cartRule->product_restriction` in `setProductRestrictions` so you don't need to call `updateRestrictionProperty` after even if it worked as expected

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]:
```suggestion
```
The field update is already handled in `setProductRestrictions`

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]:
```suggestion
```
I believe this is already handled inside `setCartRuleRestrictions`

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php]: What does ID stand for? IS it productId, categoryId? Or can it be used for different entities?

File: src/PrestaShopBundle/Resources/config/services/adapter/cart_rule.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/cart_rule.yml]: Can be cleaned I think

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature]: Since the refacto PR has been merged can you fix this part in this PR?

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (zuk3975) [reply src/Adapter/CartRule/Repository/CartRuleRepository.php]: yep, it can be category,product,attribute,supplier or manufacturer. Depends on $restriction->type

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (zuk3975) [reply src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: felt like a good shortcut :smile: I can get rid of it, at some "theoretical level" you are right :smile:

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (zuk3975) [reply src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: yeah I mixed up 2 ideas that I had and forgot to clean half of it :bow:

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php]: this "product" step can be removed. But if its the only thing to do then it shouldn't block this PR (I'll fix in upcoming PR)

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php]: removed :heavy_check_mark:

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [reply src/Adapter/CartRule/Repository/CartRuleRepository.php]: Can you please replace this public field with a getter instead (and same in other commands, although it can be done in a dedicated PR for others) This one is a bit too generic, so maybe something like:

```suggestion
                foreach ($restrictionRule->getEntityIds() as $entityId) {
```

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]:
```suggestion
        $cartRule = $this->cartRuleRepository->get($command->getCartRuleId());
```

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php]:
```suggestion
            foreach ($restrictionRuleGroup->getRestrictionRules() as $restrictionRule) {
```

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php]:
```suggestion
                    ->setParameter('type', $restrictionRule->getType())
```

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [reply src/Adapter/CartRule/Repository/CartRuleRepository.php]: Or `getRestrictedEntityIds` or `getRestrictedIds` I'm not sure

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php:178]: Seems like this query is quite similar with the previous one, they are both based on the same condition `'crprg.id_cart_rule = :cartRuleId'` and get data from related tables that are part of the joins in both queries

Isn't it possible to perform only one query and get all the needed data in one go?

File: src/Adapter/CartRule/Repository/CartRuleRepository.php
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php:194]: SAme question as the comment before, maybe even these three queries can be merged no?

File: src/Core/Domain/CartRule/Command/SetCartRuleRestrictionsCommand.php
Comment (jolelievre) [inline src/Core/Domain/CartRule/Command/SetCartRuleRestrictionsCommand.php]: Should be replaced by a protected readonly and a getter instead

File: src/Core/Domain/CartRule/QueryResult/CartRuleRestrictionsForEditing.php
Comment (jolelievre) [inline src/Core/Domain/CartRule/QueryResult/CartRuleRestrictionsForEditing.php]: SAme here, we should use getters instead

File: src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRule.php
Comment (jolelievre) [inline src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRule.php]: To replace with protected fields and getters

File: src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRuleGroup.php
Comment (jolelievre) [inline src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRuleGroup.php]: SAme here to replace with getters

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php]:
```suggestion
            ->getProductRestrictionRuleGroups()
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php]:
```suggestion
                $actualGroup->getRequiredQuantityInCart(),
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php]:
```suggestion
                $actualGroup->getRestrictionRules(),
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php:176]: Shouldn't `$this->restrictionCommandsByReference` also be cleaned once it's saved? To make sure no values from previous scenarios can remain in case a new scenario starts on the same cart rule

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature:90]:
```suggestion
    And cart rule rule_50_percent should have no product restriction rules
    And cart rule rule_70_percent should have no product restriction rules
```
Third one should be checked as well no?

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature:83]: Should you clean them in case some additional scenarios are added? Like you did in the product restrictions feature file

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature:83]:
```suggestion
    And I clear all product restrictions for cart rule rule_free_shipping_1
    And I save all the restrictions for cart rule rule_free_shipping_1
    And I clear all product restrictions for cart rule rule_50_percent
    And I save all the restrictions for cart rule rule_50_percent
    And I clear all product restrictions for cart rule rule_70_percent
    And I save all the restrictions for cart rule rule_70_percent
    And cart rule "rule_free_shipping_1" should have the following properties:
```

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature]:
```suggestion
  Scenario: Restrict cart rule products by defining supplier matching rules
```

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature:284]: SHould we add some scenarios that edit the existing restrictions after initial one have been added? I believe since we're dealing with new cart rules and new group in each scenario you do edit them once but never twice 

And especially we should test that we can remove the previous rules, even remove them all and check that no restrictions are set anymore

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature:284]: added couple steps here. Should cover it.
https://github.com/PrestaShop/PrestaShop/pull/32556/commits/781363bf56e65d1d1474debf05563de354f0dbe2

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php]: In this endpoint, you execute the command right away so you shouldn't rely on the tmp saved command here because it could already contain some values inconsistent ith what the endpoint aims at doing

I think it would be safer to create a new dedicated command, set the invalid value and execute it, additionally, the persisted command could even be cleared.

TLDR; `getRestrictionsCommand` should ONLY be used in the endpoints that need stateful steps because of their complexity, this one can remain independent and stateless

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

**Review by boherm — APPROVED** [review]
LGTM, thanks @zuk3975!

**Pull request merged**
